Follow API

akshays2112 opened the pull request:
As per Daniel's email request to me, I have changed to UriBuilder.  This also made the method I had made to build the comma-separated string from a list obsolete so that is also removed.

Thanks for your patience and time.

Akshay Srinivasan

akshays2112 committed:
Added Follow Api one call

Added the Follow Api call for checking by artist id or user id if the current logged in user is following them.

akshays2112 committed:
Changed to UriBuilder.  Tested code out with my main project and it works.

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -0,0 +1,73 @@
+        #region GetFollowingContains
+
+        /// <summary>
+        /// Check if current user follows artists or users.
+        /// </summary>
+        /// <param name="username"></param>
+        /// <param name="type">The type, either artist or user, the current user follows.</param>
+        /// <param name="ids">A comma-separated list of the artist or the user Spotify IDs to check. A maximum of 50 IDs can be sent in one request.</param>
```
Please copy the text from the Spotify reference so that it closely matches their docs.

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -0,0 +1,73 @@
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/follow/check-current-user-follows/
+        /// </remarks>
+        public async Task<List<bool>> GetFollowingContains(
```
* Let's simplify the return type to `string[]`. I don't think we need the overhead of a List here 🙂
* Please add Generic versions of all methods so that users can provide their own return type.
* Let's split this into two methods: `CheckCurrentUserFollowsArtists` and `CheckCurrentUserFollowsUsers`. I think this is easier for users. And then we can hard code the type and don't need another Enum 🙂

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -0,0 +1,73 @@
+        public async Task<List<bool>> GetFollowingContains(
+            string username,
+            FollowApi.ContainsTypes type,
+            List<string> ids,
```
Please use `string[]` here. This library's preference is for simple types wherever possible (I will document this in CONTRIB)

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -0,0 +1,73 @@
+        /// https://developer.spotify.com/documentation/web-api/reference/follow/check-current-user-follows/
+        /// </remarks>
+        public async Task<List<bool>> GetFollowingContains(
+            string username,
```
I don't think this param is needed?

DanielLarsenNZ replied:
@akshays2112 - The `username` param is not used because this endpoint is for the current user (me). You can remove it.

DanielLarsenNZ reviewed (commented):
Nice work @akshays2112 keep it up! 😁 I have made some comments in code. Let me know if any questions about anything. And thanks for your help!

akshays2112 committed:
Changes as per Daniel's comments:
https://github.com/Ringobot/SpotifyApi.NetCore/pull/29

akshays2112 committed:
Changed username param comment.

akshays2112 commented:
Hi Daniel,

I think I got all the changes done and committed. Also has been tested in my Spotify Blazor Server-side app.

Thanks for your time and patience.

Best Regards,

Akshay Srinivasan

akshays2112 commented:
Hi Daniel,

 

Just saw this email about 1 hour back approximately.  I have done all the changes and committed it.

 

Tell me if I need to do anything else.

 

I am new to all this so I documented Spotify UserName because it will help noobs like me. It confused me when I was trying to figure it out. Kept putting my login address.

 

Thanks for your patience and time.

 

Best Regards,

 

Akshay Srinivasan

 

From: Daniel Larsen <notifications@github.com> 
Sent: Wednesday, June 24, 2020 8:00 AM
To: Ringobot/SpotifyApi.NetCore <SpotifyApi.NetCore@noreply.github.com>
Cc: Akshay Srinivasan <akshays2112@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [Ringobot/SpotifyApi.NetCore] Follow API (#29)

 

@DanielLarsenNZ commented on this pull request.

Nice work @akshays2112 <https://github.com/akshays2112>  keep it up! 😁 I have made some comments in code. Let me know if any questions about anything. And thanks for your help!

  _____  

In src/SpotifyApi.NetCore/FollowApi.cs <https://github.com/Ringobot/SpotifyApi.NetCore/pull/29#discussion_r444607021> :

> +        /// Check if current user follows artists or users.
 
+        /// </summary>
 
+        /// <param name="username"></param>
 
+        /// <param name="type">The type, either artist or user, the current user follows.</param>
 
+        /// <param name="ids">A comma-separated list of the artist or the user Spotify IDs to check. A maximum of 50 IDs can be sent in one request.</param>
 

Please copy the text from the Spotify reference so that it closely matches their docs.

  _____  

In src/SpotifyApi.NetCore/FollowApi.cs <https://github.com/Ringobot/SpotifyApi.NetCore/pull/29#discussion_r444608861> :

> +        #endregion
 
+        protected internal virtual ISearchApi SearchApi { get; set; }
 
+
 
+        #region GetFollowingContains
 
+
 
+        /// <summary>
 
+        /// Check if current user follows artists or users.
 
+        /// </summary>
 
+        /// <param name="username"></param>
 
+        /// <param name="type">The type, either artist or user, the current user follows.</param>
 
+        /// <param name="ids">A comma-separated list of the artist or the user Spotify IDs to check. A maximum of 50 IDs can be sent in one request.</param>
 
+        /// <returns>List<bool></returns>
 
+        /// <remarks>
 
+        /// https://developer.spotify.com/documentation/web-api/reference/follow/check-current-user-follows/
 
+        /// </remarks>
 
+        public async Task<List<bool>> GetFollowingContains(
 

*	Let's simplify the return type to string[]. I don't think we need the overhead of a List here 🙂
*	Please add Generic versions of all methods so that users can provide their own return type.
*	Let's split this into two methods: CheckCurrentUserFollowsArtists and CheckCurrentUserFollowsUsers. I think this is easier for users. And then we can hard code the type and don't need another Enum 🙂

  _____  

In src/SpotifyApi.NetCore/FollowApi.cs <https://github.com/Ringobot/SpotifyApi.NetCore/pull/29#discussion_r444609187> :

> +        #region GetFollowingContains
 
+
 
+        /// <summary>
 
+        /// Check if current user follows artists or users.
 
+        /// </summary>
 
+        /// <param name="username"></param>
 
+        /// <param name="type">The type, either artist or user, the current user follows.</param>
 
+        /// <param name="ids">A comma-separated list of the artist or the user Spotify IDs to check. A maximum of 50 IDs can be sent in one request.</param>
 
+        /// <returns>List<bool></returns>
 
+        /// <remarks>
 
+        /// https://developer.spotify.com/documentation/web-api/reference/follow/check-current-user-follows/
 
+        /// </remarks>
 
+        public async Task<List<bool>> GetFollowingContains(
 
+            string username,
 
+            FollowApi.ContainsTypes type,
 
+            List<string> ids,
 

Please use string[] here. This library's preference is for simple types wherever possible (I will document this in CONTRIB)

  _____  

In src/SpotifyApi.NetCore/FollowApi.cs <https://github.com/Ringobot/SpotifyApi.NetCore/pull/29#discussion_r444609459> :

> +        protected internal virtual ISearchApi SearchApi { get; set; }
 
+
 
+        #region GetFollowingContains
 
+
 
+        /// <summary>
 
+        /// Check if current user follows artists or users.
 
+        /// </summary>
 
+        /// <param name="username"></param>
 
+        /// <param name="type">The type, either artist or user, the current user follows.</param>
 
+        /// <param name="ids">A comma-separated list of the artist or the user Spotify IDs to check. A maximum of 50 IDs can be sent in one request.</param>
 
+        /// <returns>List<bool></returns>
 
+        /// <remarks>
 
+        /// https://developer.spotify.com/documentation/web-api/reference/follow/check-current-user-follows/
 
+        /// </remarks>
 
+        public async Task<List<bool>> GetFollowingContains(
 
+            string username,
 

I don't think this param is needed?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/Ringobot/SpotifyApi.NetCore/pull/29#pullrequestreview-436279362> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AIRJIGT74UK7SXGG3CUGNH3RYFQKXANCNFSM4OF2YZ4A> .  <https://github.com/notifications/beacon/AIRJIGSOPC6AR3P2WAQWUILRYFQKXA5CNFSM4OF2YZ4KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGODIARQQQ.gif>

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -0,0 +1,119 @@
+            if (ids?.Length < 1 || ids?.Length > 50) throw new ArgumentNullException("ids");
+
+            var builder = new UriBuilder($"{BaseUrl}/me/following/contains");
+            builder.AppendToQueryAsCsv("type", new string[] { "artist" });
```
You can use the `AppendToQuery` extension method here, i.e. 

```
builder.AppendToQuery("type", "artist");
```

DanielLarsenNZ commented on src/SpotifyApi.NetCore/FollowApi.cs:
```diff
@@ -0,0 +1,119 @@
+            if (ids?.Length < 1 || ids?.Length > 50) throw new ArgumentNullException("ids");
+
+            var builder = new UriBuilder($"{BaseUrl}/me/following/contains");
+            builder.AppendToQueryAsCsv("type", new string[] { "user" });
```
Use `AppendToQuery()` here

DanielLarsenNZ reviewed (commented):
Almost there! 😃 Couple more comments below.

Next I need you to write a couple of Integration tests. Have a look at `ArtistsApiTests.cs` and `BrowseApiTests.cs` for examples. I will update CONTRIB with instructions on how to setup your environment for running Integration tests

DanielLarsenNZ commented:
I have documented how to setup the dev environment for Integration Tests now. See [CONTRIB.md](https://github.com/Ringobot/SpotifyApi.NetCore/blob/master/CONTRIB.md)

akshays2112 committed:
Changes as per Daniel's request.  Figured out how Unit Tests work. Problems with Authentication to do with bearer/access token fixed in my own way. Daniel needs to look into why I had to do this I think. Created the UnitTests and all passed. Changed my App to work the changes and ran it and worked. So end to end it works.

akshays2112 commented:
Hi Daniel,

I have done everything. Including the UnitTests.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

P.S. - Sorry took so much time busy with housework and taking care of old parents. Covid-19 problems :P

akshays2112 committed:
Trying to fix manually the conflict.

akshays2112 commented:
Hi Daniel,

Sorry about that problem. I fixed it manually. I tried to look through the interface but it was not obvious how to resolve the fact that my file in the fork is not up to date with yours. The conflict is gone now as I fixed it manually.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

DanielLarsenNZ reviewed (approved):
Beautiful! 🤩 👍 :shipit:

DanielLarsenNZ commented:
Now released in 3.2.1

    Install-Package SpotifyApi.NetCore -Version 3.2.1
